File: commonrv/src/main/java/com/even/commonrv/adapter/BaseViewHolder.java
package com.even.commonrv.adapter;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.util.SparseArray;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;

/**
 * @author Created by Even on 2019/2/21
 * Emial: emailtopan@163.com
 */
public class BaseViewHolder extends RecyclerView.ViewHolder {
    public View itemView;
    /**
     * 存放View
     */
    SparseArray<View> views;

    public BaseViewHolder(@NonNull View itemView) {
        super(itemView);
        this.itemView = itemView;
        views = new SparseArray<>();
    }

    /**
     * 通过Id获取到view
     *
     * @param viewId
     * @param <T>
     * @return BaseViewHolder
     */
    public <T extends View> T getView(int viewId) {
        View view = views.get(viewId);
        if (view == null) {
            view = itemView.findViewById(viewId);
            views.put(viewId, view);
        }
        return (T) view;
    }

    /**
     * 返回viewHolder给adapter使用
     *
     * @param context
     * @param parent
     * @param layoutId
     * @param <T>
     * @return BaseViewHolder
     */
    public static <T extends BaseViewHolder> T getHolder(Context context, ViewGroup parent, int layoutId) {
        return (T) new BaseViewHolder(LayoutInflater.from(context).inflate(layoutId, parent, false));
    }

    /**
     * textView设置值
     *
     * @param viewId
     * @param text
     * @return BaseViewHolder
     */
    public BaseViewHolder setText(int viewId, String text) {
        TextView view = getView(viewId);
        if (view != null) {
            view.setText(text);
        }
        return this;
    }

    /**
     * 设置文字大小
     *
     * @param viewId
     * @param textSize
     * @return  BaseViewHolder
     */
    public BaseViewHolder setTextSize(int viewId, int textSize) {
        TextView view = getView(viewId);
        view.setTextSize(TypedValue.COMPLEX_UNIT_PX, textSize);
        return this;
    }

    /**
     * 设置text View提示文字
     *
     * @param viewId
     * @param hintText
     * @return BaseViewHolder
     */
    public BaseViewHolder setTextHint(int viewId, String hintText) {
        TextView view = getView(viewId);
        view.setHint(hintText);
        return this;
    }

    /**
     * 设置字符串颜色
     */
    public BaseViewHolder setTextColor(int viewId, int textColor) {
        TextView view = getView(viewId);
        view.setTextColor(textColor);
        return this;
    }

    /**
     * 设置View是否被选中
     *
     * @param viewId
     * @param isCheck
     * @return BaseViewHolder
     */
    public BaseViewHolder setViewSelected(int viewId, boolean isCheck) {
        View view = getView(viewId);
        view.setSelected(isCheck);
        return this;
    }

    /**
     * 设置view的点击事件
     *
     * @param viewId
     * @param listener
     * @return BaseViewHolder
     */
    public BaseViewHolder setClickListener(int viewId, View.OnClickListener listener) {
        View view = getView(viewId);
        view.setOnClickListener(listener);
        return this;
    }

    /**
     * 设置view的长按事件
     *
     * @param viewId
     * @param longClickListener
     * @return BaseViewHolder
     */
    public BaseViewHolder setLongClickListener(int viewId, View.OnLongClickListener longClickListener) {
        View view = getView(viewId);
        view.setOnLongClickListener(longClickListener);
        return this;
    }

    /**
     * 设置是否可见
     *
     * @param viewId
     * @param visibility
     * @return BaseViewHolder
     */
    public BaseViewHolder setViewVisibility(int viewId, int visibility) {
        View view = getView(viewId);
        view.setVisibility(visibility);
        return this;
    }

    /**
     * 设置View的背景颜色
     */
    public BaseViewHolder setViewBgColor(int viewId, int colorId) {
        View view = getView(viewId);
        view.setBackgroundColor(colorId);
        return this;
    }

    /**
     * 设置资源图片
     */
    public BaseViewHolder setImageResource(int viewId, int resId) {
        ImageView iv = getView(viewId);
        iv.setImageResource(resId);
        return this;
    }

    /**
     * 设置View的背景图片
     *
     * @param viewId
     * @param resId
     * @return BaseViewHolder
     */
    public BaseViewHolder setViewBgResource(int viewId, int resId) {
        View view = getView(viewId);
        view.setBackgroundResource(resId);
        return this;
    }

    /**
     * 设置bitmap图片
     */
    public BaseViewHolder setImageBitmap(int viewId, Bitmap bitmap) {
        ImageView iv = getView(viewId);
        iv.setImageBitmap(bitmap);
        return this;
    }

    /**
     * 为linearLayout添加子View
     *
     * @param viewId
     * @param childView
     * @return BaseViewHolder
     */
    public BaseViewHolder addChildView(int viewId, View childView) {
        LinearLayout view = getView(viewId);
        view.addView(childView);
        return this;
    }

    /**
     * 设置网络图片并显示默认图片
     *
     * @param viewId
     * @param url
     * @param defaultImage
     * @return BaseViewHolder
     */
    public BaseViewHolder setImageByUrl(int viewId, String url, int defaultImage) {
        ImageView view = getView(viewId);
        if (!((Activity) view.getContext()).isFinishing()) {
            Glide.with(view.getContext()).load(url).apply(new RequestOptions().placeholder(defaultImage)).into(view);
        }
        return this;
    }

}

File: commonrv/src/main/java/com/even/commonrv/adapter/BaseViewHolder.kt
package com.even.commonrv.adapter

import android.app.Activity
import android.graphics.Bitmap
import android.util.SparseArray
import android.util.TypedValue
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions

/**
 * @author Created by Even on 2019/2/21
 * Email: emailtopan@163.com
 */
class BaseViewHolder(itemView: View) : ViewHolder(itemView) {

    /**
     * 存放View
     */
    private var mItemViews: SparseArray<View?> = SparseArray()

    /**
     * 通过Id获取到view
     *
     * @param viewId
     */
    fun <T : View?> getView(viewId: Int): T? {
        var view = mItemViews[viewId]
        if (view == null) {
            view = itemView.findViewById(viewId)
            mItemViews.put(viewId, view)
        }
        return view as T
    }

    /**
     * textView设置值
     */
    fun setText(viewId: Int, text: CharSequence): BaseViewHolder {
        getView<TextView>(viewId)?.text = text
        return this
    }

    /**
     * 设置文字大小
     */
    fun setTextSize(viewId: Int, textSize: Float): BaseViewHolder {
        getView<TextView>(viewId)?.setTextSize(TypedValue.COMPLEX_UNIT_SP, textSize)
        return this
    }

    /**
     * 设置text View提示文字
     */
    fun setTextHint(viewId: Int, hintText: String): BaseViewHolder {
        getView<TextView>(viewId)?.hint = hintText
        return this
    }

    /**
     * 设置字符串颜色
     */
    fun setTextColor(viewId: Int, colorId: Int): BaseViewHolder {
        getView<TextView>(viewId)?.setTextColor(colorId)
        return this
    }

    /**
     * 设置View是否被选中
     */
    fun setViewSelected(viewId: Int, isSelected: Boolean): BaseViewHolder {
        getView<View>(viewId)?.isSelected = isSelected
        return this
    }

    /**
     * 设置view的点击事件
     */
    fun setClickListener(viewId: Int, listener: View.OnClickListener): BaseViewHolder {
        getView<View>(viewId)?.setOnClickListener(listener)
        return this
    }

    /**
     * 设置view的长按事件
     */
    fun setLongClickListener(viewId: Int, longClickListener: OnLongClickListener): BaseViewHolder {
        getView<View>(viewId)?.setOnLongClickListener(longClickListener)
        return this
    }

    /**
     * 设置是否可见
     */
    fun setViewVisibility(viewId: Int, visibility: Int): BaseViewHolder {
        getView<View>(viewId)?.visibility = visibility
        return this
    }

    /**
     * 设置View的背景颜色
     */
    fun setViewBackgroundColor(viewId: Int, colorId: Int): BaseViewHolder {
        getView<View>(viewId)?.setBackgroundColor(colorId)
        return this
    }

    /**
     * 设置资源图片
     */
    fun setImageResource(viewId: Int, resId: Int): BaseViewHolder {
        getView<ImageView>(viewId)?.setImageResource(resId)
        return this
    }

    /**
     * 设置View的背景图片
     */
    fun setViewBackgroundResource(viewId: Int, resId: Int): BaseViewHolder {
        getView<View>(viewId)?.setBackgroundResource(resId)
        return this
    }

    /**
     * 设置bitmap图片
     */
    fun setImageBitmap(viewId: Int, bitmap: Bitmap?): BaseViewHolder {
        getView<ImageView>(viewId)?.setImageBitmap(bitmap)
        return this
    }

    /**
     * GroupView添加子View
     */
    fun addChildView(viewId: Int, childView: View): BaseViewHolder {
        getView<ViewGroup>(viewId)?.addView(childView)
        return this
    }

    /**
     * 设置网络图片并显示默认图片
     */
    fun setImageByUrl(viewId: Int, url: String, defaultImage: Int): BaseViewHolder {
        val view = getView<ImageView>(viewId)
        if (!(view?.context as Activity).isFinishing) {
            Glide.with(view.context).load(url).apply(RequestOptions().placeholder(defaultImage))
                .into(view)
        }
        return this
    }

    /**
     * 设置网络图片
     */
    fun setImageByUrl(viewId: Int, url: String): BaseViewHolder {
        val view = getView<ImageView>(viewId)
        if (!(view?.context as Activity).isFinishing) {
            Glide.with(view.context).load(url).into(view)
        }
        return this
    }
}